File: website_back/src/main/java/com/baronesa/website/entity/Theme.java
package com.baronesa.website.entity;

import com.baronesa.website.enums.ThemeStatus;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@Entity
@Table(name = "theme")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Theme {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(name = "base_theme_id")
    private Long baseThemeId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private ThemeStatus status = ThemeStatus.DRAFT;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, String> tokens;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> assets;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> content;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "tenant_id", nullable = false)
    private String tenantId;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: website_front/src/types/theme.ts
// src/types/theme.ts
export interface Theme {
  id: number;
  name: string;
  baseThemeId?: number;
  status: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  tokens: Record<string, string>;
  assets: Record<string, any>;
  content: Record<string, any>;
  tenantId: string;
  createdAt: string;
  updatedAt: string;
}

export interface ThemeDTO {
  name: string;
  baseThemeId?: number;
  status?: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED';
  tokens?: Record<string, string>;
  assets?: Record<string, any>;
  content?: Record<string, any>;
  tenantId: string;
}

export interface ThemeScheduleDTO {
  themeId: number;
  validFrom?: string;
  validTo?: string;
  priority?: number;
}
